### PR TITLE
65 classes parameters through env variables

### LINKED ISSUE #65: Classes parameters through env variables
### What is the expected enhancement?

In order to smoothly integrate purple caffeine with other system it might be beneficial to be able to create object without exactly specifing all constructor parameters, but provide them though env variables.

Let me explain. Instead of this

```python
class Trial:
  def __init__(self, name: str, storage: Optional[BaseStorage] = None):
    self.name = name
    self.storage = storage
```

have something like this

```python
class Trial:
  def __init__(self, name: Optional[str] = None, storage: Optional[BaseStorage] = None):
    self.name = name or os.environ.get("PURPLE_CAFFEINE_TRIAL_NAME")
    if self.name is None:
      raise PurplecaffeineException("Please specify name or trial or configure it using env variables")
    if storage is None:
      storage_type = os.environ.get("PURPLE_CAFFEINE_STORAGE_CLASS", "LocalStorage")
      storage_mapping: Dict[str, BaseStorage] = {
        "LocalStorage": LocalStorage,
        "S3Storage": S3Storage,
        "ApiStorage": ApiStorage
      }
      self.storage = storage_type.get(storage_type)()
    else:
      self.storage = storage
```

Now if we have all env variable properly set we can just write
```python
with Trial() as trial:
    trial.add_...
```

We need to have the same behavior with all parameters for Trial, LocalStorage, S3Storage, ApiStorage.

### PR DESCRIPTION
**Pull request by AnsahMohammad**
### Summary

Added environment variables for :
- [x] Trial class
- [x] Api Storage class
- [x] Local storage class
- [x] S3 Storage class

### CODE REVIEW

File: client/purplecaffeine/core.py [hunk 1]
```diff
@@ -67,18 +67,40 @@ def __init__(
...
+                "LocalStorage": LocalStorage,
+                "S3Storage": S3Storage,
+                "ApiStorage": ApiStorage
+            }
+            self.storage = storage_mapping.get(storage_type)()
+        else:
+            self.storage = storage
+
+        self.description = description or os.environ.get("PURPLE_CAFFEINE_TRIAL_DESCRIPTION", "")
+        self.metrics = metrics or os.environ.get("PURPLE_CAFFEINE_TRIAL_METRICS", "").split(",")
```
Comment (IceKhan13) [inline client/purplecaffeine/core.py]: I do not think we need metrics, parameters, circuit, operators, artifacts, texts, array to be from env vars.

Everything else is good!

Comment (AnsahMohammad) [reply client/purplecaffeine/core.py]: What about `tags`? should it be allowed to be taken from env variables?

Comment (IceKhan13) [reply client/purplecaffeine/core.py]: yes, tags should be possible to set from env

**Comment by AnsahMohammad** [comment]
@IceKhan13, what should be done with the Access Key is not available for S3 storage in the `unit test & coverage` workflow?

**Comment by IceKhan13** [comment]
Error is saying 

```
======================================================================
ERROR: test_save_get_list_s3_storage (tests.test_storage.TestStorage)
Test of S3Storage object.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/purplecaffeine/purplecaffeine/client/tests/test_storage.py", line 74, in test_save_get_list_s3_storage
    s3_storage = S3Storage(
  File "/home/runner/work/purplecaffeine/purplecaffeine/client/purplecaffeine/core.py", line 564, in __init__
    raise PurpleCaffeineException(
purplecaffeine.exception.PurpleCaffeineException: Please specify Access key of S3 Bucket or configure it using env variables
```

which means error from https://github.com/IceKhan13/purplecaffeine/blob/faf5d439371c32d81df67201130931d84069f620/client/purplecaffeine/core.py#L562-L566 is raised 

I think it might be a problem since we are passing access_key, it's just empty string https://github.com/IceKhan13/purplecaffeine/blob/2239c0308020bc819990982d59eafe9f761a10dc/client/tests/test_storage.py#L74-L79

Therefore we might need to check access key not like this `access or os.env...` but something like:
```python
self.access_key = os.environ.get(...)
if access_key is not None:
    self.access_key = access_key

...

if self.access_key is None or self.secret_access_key is None:
    raise ...
```

**Comment by AnsahMohammad** [comment]
@IceKhan13 yepp that was the issue :smile:, solved now :tada:

File: client/purplecaffeine/core.py [hunk 2]
```diff
@@ -510,15 +554,34 @@ def __init__(
...
+            )
+        self.directory = directory or os.environ.get("PURPLE_CAFFEINE_S3_DIRECTORY")
+        self.endpoint_url = endpoint_url or os.environ.get(
+            "PURPLE_CAFFEINE_S3_ENDPOINT"
+        )
         client_s3 = boto3.client(
             "s3",
-            aws_access_key_id=access_key,
+            aws_access_key_id=self.access_key,
             aws_secret_access_key=secret_access_key,
```
Comment (IceKhan13) [inline client/purplecaffeine/core.py]:
```suggestion
            aws_secret_access_key=self.secret_access_key,
```

**Review by IceKhan13 — APPROVED** [review]
ok, now we can deploy this to any cloud service with cridentails, so users do not need to worry about copy-pasting auth details :) 

Great job!

**Pull request merged**